### PR TITLE
migrate slot to use types from typingref

### PR DESCRIPTION
**Pull request by xiangxw**
https://github.com/nrbnlulu/qtgql/blob/main/qtgql/slot.py#L35 slot implementation is
touching type hints by hand, migrate to use TypeHinter
issue #3

### CODE REVIEW

**Comment by nrbnlulu** [comment]
Hi can you add RELEASE.md?

* Release type: major, minor, or patch
* Description of the changes, to be used as the changelog entry

Example:

    Release type: patch

    Add function to update version strings in multiple files.

File: qtgql/slot.py [hunk 1]
```diff
@@ -1,52 +1,50 @@
...
+
 def slot(func):
     def wrapper(func) -> func:
         anots: dict = func.__annotations__
         return_ = anots.pop("return", None)
         if return_ is Any:
             return_ = "QVariant"
-        args = list(anots.values())
+        args = [TypeHinter.from_annotations(th) for th in anots.values()]
         stripped_optionals = get_concretes(get_optional_args(args))
```
Comment (nrbnlulu) [inline qtgql/slot.py:39]:
```suggestion
        stripped_optionals = [th.of_type[0] for th in args if th.is_optional()]
```

File: qtgql/slot.py [hunk 2]
```diff
@@ -1,52 +1,50 @@
...
         concretes = get_concretes(args)
```
Comment (nrbnlulu) [inline qtgql/slot.py:40]:
```suggestion
        concretes = [th for th in args if not th.is_optional()]
```

File: qtgql/slot.py [hunk 3]
```diff
@@ -1,52 +1,50 @@
...
-from typing import Any, Optional, get_args
+from typing import Any
 
 from PySide6 import QtCore as qtc
 
-from qtgql.typingref import is_optional
+from qtgql.typingref import TypeHinter
 
 
-def get_optional_args(annotations: list[Any]) -> list[Any]:
```
Comment (nrbnlulu) [inline qtgql/slot.py:9]: redundant

File: qtgql/slot.py [hunk 4]
```diff
@@ -1,52 +1,50 @@
...
-    return [get_args(arg)[0] for arg in annotations if is_optional(arg)]
+def get_optional_args(annotations: list[TypeHinter]) -> list[TypeHinter]:
+    return [arg.of_type[0] for arg in annotations if arg.is_optional()]
 
 
-def get_concrete(annotation: Any) -> Optional[type]:
```
Comment (nrbnlulu) [inline qtgql/slot.py:13]: same

File: qtgql/slot.py [hunk 5]
```diff
@@ -1,52 +1,50 @@
...
-    return getattr(annotation, "__origin__", None)
-
-
-def get_concretes(annotations: list) -> list:
```
Comment (nrbnlulu) [inline qtgql/slot.py:17]: same

File: qtgql/slot.py [hunk 6]
```diff
@@ -1,52 +1,50 @@
...
             continue
-        if concrete := get_concrete(item):
-            ret.append(concrete)
-            annotations.remove(item)
+        ret.append(item)
     return ret + annotations
 
 
+def get_combos(stripped_optionals: list[TypeHinter], concretes: list[TypeHinter]):
+    required_annotations = tuple([th.type for th in concretes])
```
Comment (nrbnlulu) [inline qtgql/slot.py:24]:
```suggestion
    required_annotations = tuple([th.as_annotation() for th in concretes])
```

**Review by nrbnlulu — CHANGES_REQUESTED** [review]
Hi!
- There are some redundant code there. 
- Please add a release file. 
- Please add a test for `__future__` annotations

**Comment by xiangxw** [comment]
I will do these things mentioned above immediately

File: qtgql/slot.py (hunk 6 above)
Comment (xiangxw) [reply qtgql/slot.py:24]: pytest will fail if changed as suggested.
ImportError while loading conftest '/home/xiang/qtgql/tests/test_itemsystem/conftest.py'.
test_itemsystem/conftest.py:6: in <module>
    from qtgql.itemsystem import GenericModel, get_base_type, role
../qtgql/itemsystem/__init__.py:1: in <module>
    from .core import GenericModel, RoleDoesNotExist, asdict, get_base_type, role
../qtgql/itemsystem/core.py:164: in <module>
    class GenericModel(Generic[_TBaseType], QAbstractListModel):
../qtgql/itemsystem/core.py:269: in GenericModel
    def pop(self, index: Optional[int] = None) -> None:
../qtgql/slot.py:36: in slot
    return wrapper(func)
../qtgql/slot.py:34: in wrapper
    return qtc.Slot(*[th.as_annotation() for th in concretes], result=return_)(func)
../qtgql/slot.py:34: in <listcomp>
    return qtc.Slot(*[th.as_annotation() for th in concretes], result=return_)(func)
../qtgql/typingref.py:66: in as_annotation
    assert object_map, "can't evaluate forward refs without object_map."
E   AssertionError: can't evaluate forward refs without object_map.

Comment (nrbnlulu) [reply qtgql/slot.py:24]: make sure you are not using `from __future__ import annotations` This should be a standalone test.

Comment (nrbnlulu) [reply qtgql/slot.py:24]: also please push you commit here

Comment (xiangxw) [reply qtgql/slot.py:24]: yes, I run `pytest test_slot.py`, it failed.

Comment (xiangxw) [reply qtgql/slot.py:24]: The code in `main` branch use `getattr(annotation, "__origin__", None)`, so I think there is no need to use `as_annotation()`

**Comment by xiangxw** [comment]
`Please add a test for __future__ annotations`, seems have nothing to do with this issue #3 ?

**Comment by nrbnlulu** [comment]
fair point.

**Comment by nrbnlulu** [comment]
just add a release file and we I think its ready.

**Pull request merged**